internal/jobcontainers/process.go

```diff
@@ -0,0 +1,179 @@
+package jobcontainers
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"os/exec"
+	"sync"
+
+	"github.com/Microsoft/hcsshim/internal/cow"
+	"github.com/Microsoft/hcsshim/internal/log"
+	"golang.org/x/sys/windows"
+)
+
+// JobProcess represents a process run in a job object.
+type JobProcess struct {
+	cmd            *exec.Cmd
+	procLock       sync.Mutex
+	stdioLock      sync.Mutex
+	stdin          io.WriteCloser
+	stdout         io.ReadCloser
+	stderr         io.ReadCloser
+	waitBlock      chan struct{}
+	closedWaitOnce sync.Once
+	waitError      error
+}
+
+var _ cow.Process = &JobProcess{}
+
+func newProcess(cmd *exec.Cmd) *JobProcess {
+	return &JobProcess{
+		cmd:       cmd,
+		waitBlock: make(chan struct{}),
+	}
+}
+
+// ResizeConsole - TODO (dcantah): Implement console support with new pseudo console api
+func (p *JobProcess) ResizeConsole(ctx context.Context, width, height uint16) error {
+	return errors.New("not implemented")
+}
+
+// Stdio returns the stdio pipes of the process
+func (p *JobProcess) Stdio() (io.Writer, io.Reader, io.Reader) {
+	return p.stdin, p.stdout, p.stderr
+}
+
+// Signal sends a signal to the process.
+func (p *JobProcess) Signal(ctx context.Context, options interface{}) (bool, error) {
+	if p.exited() {
+		return false, errors.New("signal not sent. process already exited")
+	}
+
+	p.procLock.Lock()
+	defer p.procLock.Unlock()
+	if err := windows.GenerateConsoleCtrlEvent(windows.CTRL_BREAK_EVENT, uint32(p.Pid())); err != nil {
+		return false, fmt.Errorf("failed to send signal: %s", err)
+	}
+	// Process wasn't killed from the signal. Ping handles ctrl-break for example
+	// and we can't issue ctrl-c with a new process group.
+	if p.cmd.ProcessState == nil {
+		if err := p.cmd.Process.Kill(); err != nil {
+			return false, err
+		}
+	}
+	return true, nil
+}
+
+// CloseStdin closes the stdin pipe of the process.
+func (p *JobProcess) CloseStdin(ctx context.Context) error {
+	p.stdioLock.Lock()
+	defer p.stdioLock.Unlock()
+	return p.stdin.Close()
+}
+
+// Wait waits for the process to exit. If the process has already exited returns
+// the previous error (if any).
+func (p *JobProcess) Wait() error {
+	<-p.waitBlock
+	return p.waitError
+}
+
+// Start starts the job object process
+func (p *JobProcess) Start() error {
+	return p.cmd.Start()
+}
+
+// This should only be called once.
+func (p *JobProcess) waitBackground(ctx context.Context) {
+	log.G(ctx).WithField("pid", p.Pid()).Debug("waitBackground for JobProcess")
+
+	var err error
+	if p.cmd.Process == nil {
+		err = errors.New("process has not been started")
+	}
+
+	// Wait for process to get signaled/exit/terminate/.
+	err = p.cmd.Wait()
+
+	// Wait closes the stdio pipes so theres no need to later on.
+	p.stdioLock.Lock()
+	p.stdin = nil
+	p.stdout = nil
+	p.stderr = nil
+	p.stdioLock.Unlock()
+
+	p.closedWaitOnce.Do(func() {
+		p.waitError = err
+		close(p.waitBlock)
+	})
```

Comment:
I know since all the code isn't PR'd yet, I could be missing something, but in the test that uses this, we also call job.Close() afterwards and it basically has the same logic as this function. Is there a reason for that?

Reply:
I'm not following, we call job.Close after this? Where? job.Close just closes the job handle, this waits for the process to complete and will close the wait channel (not used in this PR).

Reply:
Maybe I'm also missing something 😋

Reply:
I think I understand. Waitbackground will wait before cleaning up whereas Close will forcefully close and clean up the process. Is this correct?

Reply:
jobprocess.Close frees up resources (just closes the stdin pipes as we dont have access to the actual process handle and if this is called before the process finishes it will also close the waitBlock channel) whereas waitBackground will just sit in a background thread and wait for the process to complete and go's exec.Cmd functionality will free up any resources for the process itself. This is just the same model we have for regular containers.

Reply:
I'm going to leave it as is if thats ok